mmaction/datasets/pipelines/loading.py

```diff
@@ -270,6 +270,140 @@ def _get_test_clips(self, num_frames):
+        if isinstance(aug_ratio, (int, float)):
+            self.aug_ratio = (aug_ratio, aug_ratio)
+        else:
+            assert isinstance(aug_ratio, (tuple, list))
+            assert len(aug_ratio) == 2
+            self.aug_ratio = aug_ratio
+        self.train_mode = train_mode
+        self.test_mode = test_mode
+        self.test_interval = test_interval
+
+    def _get_train_indices(self, valid_length, num_segments):
+        average_duration = (valid_length + 1) // num_segments
+        if average_duration > 0:
+            offsets = np.multiply(
+                                  list(range(num_segments)),
+                                  average_duration) + \
+                                  np.random.randint(average_duration,
+                                                    size=num_segments)
+        elif valid_length > num_segments:
+            offsets = np.sort(
+                np.random.randint(valid_length, size=num_segments))
+        else:
+            offsets = np.zeros((num_segments, ))
+
+        return offsets
+
+    def _get_val_indices(self, valid_length, num_segments):
+        if valid_length > num_segments:
+            tick = valid_length / float(num_segments)
+            offsets = np.array(
+                [int(tick / 2.0 + tick * x) for x in range(num_segments)])
+        else:
+            offsets = np.zeros((num_segments, ))
+
+        return offsets
+
+    def _get_proposal_clips(self, proposal, num_frames):
+        start_frame = proposal.start_frame + 1
+        end_frame = proposal.end_frame
+
+        duration = end_frame - start_frame + 1
+        assert duration != 0
+        valid_length = duration - self.clip_len
+
+        valid_starting = max(1,
+                             start_frame - int(duration * self.aug_ratio[0]))
+        valid_ending = min(num_frames - self.clip_len + 1,
+                           end_frame + int(duration * self.aug_ratio[1]))
+
+        valid_starting_length = start_frame - valid_starting - \
+            self.clip_len + 1
+        valid_ending_length = valid_ending - end_frame - self.clip_len + 1
+
+        starting_offsets = self._get_train_indices(
+            valid_starting_length, self.aug_segments[0]) if self.train_mode \
+            else self._get_val_indices(valid_starting_length,
+                                       self.aug_segments[0])
+        starting_offsets += valid_starting
+        course_offsets = self._get_train_indices(
+            valid_length, self.body_segments) if self.train_mode \
+            else self._get_val_indices(valid_length, self.body_segments)
+        course_offsets += start_frame
+        # yapf: disable
+        ending_offsets = self._get_train_indices(
+            valid_ending_length, self.aug_segments[1]) if self.train_mode \
+            else self._get_val_indices(valid_ending_length, self.aug_segments[1])  # noqa:E501
+        # yapf: enable
+        ending_offsets += end_frame
```

Comment:
Do not need `self.train_mode`, just using `self.test_mode` for validation is ok.
```python
if self.test_mode:
    starting_offsets = self._get_val_indices(valid_starting_length, self.aug_segments[0])
    course_offsets = self._get_val_indices(valid_length, self.body_segments)
    ending_offsets = self._get_val_indices(valid_ending_length, self.aug_segments[1])
else:
    starting_offsets = self._get_train_indices(valid_starting_length, self.aug_segments[0])
    course_offsets = self._get_train_indices(valid_length, self.body_segments)
    ending_offsets = self._get_train_indices(valid_ending_length, self.aug_segments[1])
starting_offsets += valid_starting
course_offsets += start_frame
ending_offsets += end_frame
```